ykcompile/src/lib.rs

```diff
@@ -473,160 +485,99 @@ impl<TT> TraceCompiler<TT> {
-                        );
-                    }
-                    16 => {
-                        // We could have handled this case via a generic memcpy() solution, but it
-                        // hardly seems worth the overhead for such a small copy. Especially since
-                        // pairs (tuples) of two u64s are very common in Rust: they are the result
-                        // of checked arithmetic.
-                        dynasm!(self.asm
-                            ; mov rax, [Rq(ro2.reg) + ro2.offs]
-                            ; mov [Rq(ro1.reg) + ro1.offs], rax
-                            ; mov rax, [Rq(ro2.reg) + ro2.offs + 8]
-                            ; mov [Rq(ro1.reg) + ro1.offs + 8], rax
-                        );
-                    }
-                    _ => todo!("{}", ty.size()), // FIXME: For things >16, use memcpy().
-                }
-            }
             _ => unreachable!(),
         }
-
-        // Free temporary if one was created.
-        self.free_if_temp(lloc);
-        self.free_if_temp(rloc);
+        self.free_if_temp(loc);
     }
 
-    fn mov_place_ref(&mut self, p1: &Place, p2: &Place) {
-        let (lloc, _) = self.place_to_location(p1);
-
+    /// Codegen a reference `p` and store the result in `reg`.
+    fn c_ref(&mut self, reg: u8, p: &Place) {
         // Deal with the special case `&*`, i.e. referencing a `Deref` on a reference just returns
         // the reference.
-        if let Some(pj) = p2.projection.get(0) {
+        if let Some(pj) = p.projection.get(0) {
             if matches!(pj, Projection::Deref)
-                && matches!(SIR.ty(&self.local_decls[&p2.local].ty), Ty::Ref(_))
+                && matches!(SIR.ty(&self.local_decls[&p.local].ty), Ty::Ref(_))
             {
                 // Clone the projection while removing the `Deref` from the end.
                 let mut newproj = Vec::new();
-                for p in p2.projection.iter().take(p2.projection.len() - 1) {
+                for p in p.projection.iter().take(p.projection.len() - 1) {
                     newproj.push(p.clone());
                 }
                 let np = Place {
-                    local: p2.local,
+                    local: p.local,
                     projection: newproj,
                 };
-                let (rloc, _) = self.place_to_location(&np);
-                match (lloc, rloc) {
-                    (Location::Register(reg1), Location::Register(reg2)) => {
+                let (rloc, _) = self.place_to_location(&np, false);
+                match rloc {
+                    Location::Register(reg2, _) => {
                         dynasm!(self.asm
-                            ; mov Rq(reg1), Rq(reg2)
+                            ; mov Rq(reg), Rq(reg2)
                         );
                     }
                     _ => todo!(),
                 }
+                self.free_if_temp(rloc); // ????
```

Comment:
This comment doesn't fill me with confidence lol

Reply:
Whoops. This was a marker because we wanted to double check if this statement is actually needed. Will remove.

Reply:
hah!

Reply:
I think this still needs fixing?